### PR TITLE
Allow configuration options for Postgres

### LINKED ISSUE #2872: Allow configuration options for Postgres
The current implementation of the Postgres adapter establishes a connection using the default settings. It would be helpful if we could customize the options available in the Postgres driver. As things stand, there are two sets of options that can be configured:

- The [initialization options](https://github.com/vitaly-t/pg-promise/wiki/Connection-Syntax#configuration-object) for the `pg-promise` wrapper
- The [options](https://github.com/brianc/node-postgres/blob/master/lib/defaults.js) for the `node-postgres` instance available via `pgp.pg.defaults.OPTION_NAME`

To implement this, parse server can accept them individually as command line parameters/environment variable (the existing `--databaseOptions` option and a new option). 

Alternatively, they can be combined into the existing `--databaseOptions` option with the convention that this option have a structure similar to:

```json
{
  "pgpOptions": { },
  "pgOptions": { }
}
```

Of course, the keys could have more descriptive names. If we can decide on the preferable approach and the naming convention, I could take a stab at implementing this.

### PR DESCRIPTION
**Pull request by kulshekhar**
Resolves #2872

### CODE REVIEW

**Comment by flovilmart** [comment]
@kulshekhar can you please add unit tests for the configuration parsing please? so we make sure this piece is correctly covered?

Maybe refactor into a PGConfiguration parser (that would parse the options) and then keep the initialization 'outside' the configuration parser?

**Comment by kulshekhar** [comment]
Sure. 

Just so we're on the same page, would this mean moving the newly created `getDatabaseOptionsFromURI` and `parseQueryParams` functions into a separate PGConfigurationParser file?

**Comment by flovilmart** [comment]
Yep That would be nice, so that's properly tested on itself with different URLs and options, and properly covered.

File: src/Adapters/Storage/Postgres/PostgresConfigParser.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
+const url = require('url');
+
+function getDatabaseOptionsFromURI(uri) {
+  const databaseOptions = {};
+
+  try {
```
Comment (flovilmart) [inline src/Adapters/Storage/Postgres/PostgresConfigParser.js]: Why do we surround with try/catch just to retrhow?

File: src/Adapters/Storage/Postgres/PostgresConfigParser.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+          ? decodeURIComponent(parts.slice(1).join('='))
+          : '';
+      return p;
+    }, {});
+}
+
+module.exports = {
+  parseQueryParams: parseQueryParams,
+  getDatabaseOptionsFromURI: getDatabaseOptionsFromURI
+};
```
Comment (flovilmart) [inline src/Adapters/Storage/Postgres/PostgresConfigParser.js]: Add line at the end of the file

File: src/Adapters/Storage/Postgres/ClientCreator.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
+const pgp = require('pg-promise')();
```
Comment (flovilmart) [inline src/Adapters/Storage/Postgres/ClientCreator.js]: Maybe rename file to PGClient

Comment (flovilmart) [inline src/Adapters/Storage/Postgres/ClientCreator.js]: Or PostgresClient for the sake of consistency :)

File: src/Adapters/Storage/Postgres/PostgresConfigParser.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    databaseOptions.password = authParts.length > 1 ? authParts[1] : '';
+
+    databaseOptions.ssl =
+      queryParams.ssl && queryParams.ssl.toLowerCase() === 'true' ? true : false;
+    databaseOptions.binary =
+      queryParams.binary && queryParams.binary.toLowerCase() === 'true' ? true : false;
+
+    databaseOptions.client_encoding = queryParams.client_encoding;
+    databaseOptions.application_name = queryParams.application_name;
+    databaseOptions.fallback_application_name = queryParams.fallback_application_name;
```
Comment (flovilmart) [inline src/Adapters/Storage/Postgres/PostgresConfigParser.js]: Are there additional options possible?

File: src/Adapters/Storage/Postgres/ClientCreator.js (hunk 3 above)
Comment (kulshekhar) [inline src/Adapters/Storage/Postgres/ClientCreator.js]: This isn't the client. It's just pg-promise being initialized. A call to `pgp()` is what creates the client.

Comment (flovilmart) [inline src/Adapters/Storage/Postgres/ClientCreator.js]: So PostgresClient.createClient() still makes sense :)

File: src/Adapters/Storage/Postgres/PostgresConfigParser.js (hunk 2 above)
Comment (kulshekhar) [inline src/Adapters/Storage/Postgres/PostgresConfigParser.js]: Done

File: src/Adapters/Storage/Postgres/PostgresConfigParser.js (hunk 4 above)
Comment (flovilmart) [inline src/Adapters/Storage/Postgres/PostgresConfigParser.js]: Ok good, we'll keep an eye if new options arise from that module

**Comment by lsxredrain** [comment]
nice fuction!

**Comment by lsxredrain** [comment]
name: 'error',
  length: 108,
  severity: 'ERROR',
  code: '42703',
  detail: undefined,
  hint: undefined,
  position: '33',
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'parse_relation.c',
  line: '3090',
  routine: 'errorMissingColumn' } error: column "password" does not exist
    at Connection.parseE (/data/work/psql-parser/parse-server-example/node_modules/pg/lib/connection.js:                                                              539:11)
    at Connection.parseMessage (/data/work/psql-parser/parse-server-example/node_modules/pg/lib/connecti                                                              on.js:366:17)
    at Socket.<anonymous> (/data/work/psql-parser/parse-server-example/node_modules/pg/lib/connection.js                                                              :105:22)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:543:20)
error: column "password" does not exist
    at Connection.parseE (/data/work/psql-parser/parse-server-example/node_modules/pg/lib/connection.js:                                                              539:11)
    at Connection.parseMessage (/data/work/psql-parser/parse-server-example/node_modules/pg/lib/connecti                                                              on.js:366:17)
    at Socket.<anonymous> (/data/work/psql-parser/parse-server-example/node_modules/pg/lib/connection.js                                                              :105:22)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:543:20)

**Pull request merged**